### PR TITLE
add conversion level - magnification factor, when possible

### PR DESCRIPTION
**Pull request by nicolebussola**
Address Issue #294

### CODE REVIEW

File: src/histolab/exceptions.py [hunk 1]
```diff
@@ -37,6 +37,10 @@ class LevelError(HistolabException):
     """Raised when a requested level is not available"""
 
 
+class PropertyError(HistolabException):
```
Comment (ernestoarbitrio) [inline src/histolab/exceptions.py:40]: I would call it with the `Slide` prefix considering that the docstring says :`Raised when a requested slide property is not available` . So if it is a specific Exception for a slide I would use that prefix

`SlidePropertyError`

File: src/histolab/slide.py [hunk 2]
```diff
@@ -135,6 +135,39 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+        Parameters
+        ---------
+        level : int
+            The level which magnification factor is requested, default is 0
+
+        Returns
+        -------
+        magnification factor : str
+            Magnification factor at speficied level
+        """
```
Comment (ernestoarbitrio) [inline src/histolab/slide.py:152]: Miss the 2 Raises in the docstring

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
per comments

File: docs/quickstart.rst [hunk 3]
```diff
@@ -114,6 +114,27 @@ dimensions at native magnification or at a specified level:
    Dimensions at level 1: (7500, 8496)
    Dimensions at level 2: (1875, 2124)
 
+.. note::
+    If the native magnification, *i.e.* the magnification factor used to scan the slide, is provided in the slide properties, it is also possible
```
Comment (alessiamarcolini) [inline docs/quickstart.rst:118]:
```suggestion
    If the native magnification, *i.e.,* the magnification factor used to scan the slide, is provided in the slide properties, it is also possible
```

File: src/histolab/slide.py [hunk 4]
```diff
@@ -48,6 +48,11 @@ class Slide:
         Path where the WSI is saved.
     processed_path : Union[str, pathlib.Path]
         Path where the tiles will be saved to.
+
+    Raises
+    ------
+    TypeError:
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:54]:
```suggestion
    TypeError
```

File: src/histolab/slide.py [hunk 5]
```diff
@@ -125,6 +130,11 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
         -------
         dimensions : Tuple[int, int]
             Slide dimensions at the specified level (width, height)
+
+        Raises
+        ------
+        IndexError:
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:136]:
```suggestion
        LevelError
```

File: src/histolab/slide.py [hunk 6]
```diff
@@ -135,6 +145,53 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+    def level_magnification_factor(self, level: int = 0) -> str:
+        """Return the magnification factor at the specified level.
+
+         Notice that the conversion level-magnification can be computed only
+        if the native magnification is available in the slide metadata.
+
+        Parameters
+        ---------
+        level : int
+            The level which magnification factor is requested, default is 0
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:157]:
```suggestion
            The level which magnification factor is requested, default is 0.
```

File: src/histolab/slide.py [hunk 7]
```diff
@@ -135,6 +145,53 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+
+        Returns
+        -------
+        magnification factor : str
+            Magnification factor at speficied level
+
+        Raises
+        ------
+        LevelError:
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:166]:
```suggestion
        LevelError
```

File: src/histolab/slide.py [hunk 8]
```diff
@@ -135,6 +145,53 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+            If the specified level is not available.
+        KeyError:
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:168]:
```suggestion
        SlidePropertyError
```

File: src/histolab/slide.py [hunk 9]
```diff
@@ -135,6 +145,53 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+            If the slide's native magnification is not available in the file
+            metadata.
+        """
+        level = level if level >= 0 else self._remap_level(level)
+        if level > (len(self._wsi.level_dimensions) + 1):
+            raise LevelError(
+                f"Level {level} not available. Number of available levels: "
+                f"{len(self._wsi.level_dimensions)}"
+            )
+        properties = self._wsi.properties
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:178]: I would use our wrapper
```suggestion
        properties = self.properties
```

File: src/histolab/slide.py [hunk 10]
```diff
@@ -135,6 +145,53 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+        Raises
+        ------
+        LevelError:
+            If the specified level is not available.
+        KeyError:
+            If the slide's native magnification is not available in the file
+            metadata.
+        """
+        level = level if level >= 0 else self._remap_level(level)
+        if level > (len(self._wsi.level_dimensions) + 1):
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:173]:
```suggestion
        if level in self.levels:
```

File: src/histolab/slide.py [hunk 11]
```diff
@@ -135,6 +145,53 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+            else 1
+        )
+        try:
+            level_magnification = (
+                int(properties["openslide.objective-power"]) / downsample_factor
+            )
+            return f"{level_magnification}X"
+        except KeyError:
+            raise SlidePropertyError(
+                f"Native magnification not available. Available slide properties: "
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:191]:
```suggestion
                "Native magnification not available. Available slide properties: "
```

File: src/histolab/slide.py [hunk 12]
```diff
@@ -135,6 +145,53 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+                f"{list(self._wsi.properties.keys())}"
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:192]:
```suggestion
                f"{list(self.properties.keys())}"
```

File: tests/unit/test_slide.py [hunk 13]
```diff
@@ -277,6 +278,80 @@ def but_it_raises_expection_when_level_does_not_exist(self, level, tmpdir):
...
+
+        with pytest.raises(LevelError) as err:
+            slide.level_magnification_factor(level=level)
+
+        assert isinstance(err.value, LevelError)
+        assert (
+            str(err.value)
+            == f"Level {level} not available. Number of available levels: 3"
+        )
+
```
Comment (alessiamarcolini) [inline tests/unit/test_slide.py:354]: Usually we test exceptions with unit tests to make them fast (mocking where needed)

File: tests/unit/test_slide.py [hunk 14]
```diff
@@ -277,6 +278,80 @@ def but_it_raises_expection_when_level_does_not_exist(self, level, tmpdir):
...
+                marks=pytest.mark.skipif(not on_ci(), reason="To run only on CI"),
+            ),
+        ),
+    )
+    def it_knows_its_magnification_factors(self, fixture_slide, level, expected_value):
+        slide = Slide(fixture_slide, "")
+
+        magnification_factor = slide.level_magnification_factor(level=level)
+
+        assert magnification_factor == expected_value
```
Comment (alessiamarcolini) [inline tests/unit/test_slide.py:309]: these tests should go with the integration tests as we are using real slides, and unit tests for this method are missing :)

**Review by alessiamarcolini — CHANGES_REQUESTED** [review]

File: docs/quickstart.rst (hunk 3 above)
Comment (nicolebussola) [reply docs/quickstart.rst:118]: not sure its needed

Comment (alessiamarcolini) [reply docs/quickstart.rst:118]: it depends whether we are writing British or American. I guess we always used American english

**Review by alessiamarcolini — CHANGES_REQUESTED** [review]
Now the tests that are included are ok. 

However, since the integration test `it_knows_its_magnification_factors` is basically run only on CI, I would add:
* a test case for that method to use the local `cmu-1-small-region.svs`
* two unit tests using a fake svs and mocking `properties` property to include or not `"openslide.objective-power"` and `"f"openslide.level[{level}].downsample"` to better control the flow. Let's try to have 100% coverage on the new modifications without the help of the integration tests

File: src/histolab/slide.py [hunk 15]
```diff
@@ -135,6 +145,59 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+            If the slide's native magnification is not available in the file
+            metadata.
+        """
+        level = level if level >= 0 else self._remap_level(level)
+        properties = self.properties
+
+        if level not in self.levels:
+            raise LevelError(
+                f"Level {level} not available. Number of available levels: "
+                f"{len(self._wsi.level_dimensions)}"
```
Comment (alessiamarcolini) [inline src/histolab/slide.py:178]: here you should also use `len(self.levels)`, and then you will see that the unit test `and_it_raises_an_exception_if_level_in_incorrect` will fail (as it should)

File: tests/unit/test_slide.py [hunk 16]
```diff
@@ -277,6 +277,109 @@ def but_it_raises_expection_when_level_does_not_exist(self, level, tmpdir):
...
+            (
+                {
+                    "openslide.objective-power": "20",
+                    "openslide.level[1].downsample": "4.003",
+                    "openslide.level[2].downsample": "16",
+                },
+            ),
+        ),
+    )
+    def and_it_raises_an_exception_if_level_in_incorrect(
```
Comment (alessiamarcolini) [inline tests/unit/test_slide.py:344]:
```suggestion
    def and_it_raises_an_exception_if_level_is_incorrect(
```

File: tests/unit/test_slide.py [hunk 17]
```diff
@@ -277,6 +277,109 @@ def but_it_raises_expection_when_level_does_not_exist(self, level, tmpdir):
...
+        self, request, properties, tmpdir
+    ):
+        slide, _ = base_test_slide(tmpdir, PILIMG.RGBA_COLOR_500X500_155_249_240)
+        property_mock(request, Slide, "properties", return_value=properties)
+        property_mock(request, Slide, "levels", return_value=[0, 1, 2])
+        with pytest.raises(LevelError) as err:
+            slide.level_magnification_factor(4)
+
+        assert isinstance(err.value, LevelError)
+        assert str(err.value) == "Level 4 not available. Number of available levels: 1"
```
Comment (alessiamarcolini) [inline tests/unit/test_slide.py:354]: getting back to the previous comment... here `levels` is mocked as `[0, 1, 2]`, while the error message says that there is 1 level.... obv wrong :D

**Review by alessiamarcolini — CHANGES_REQUESTED** [review]

**Review by alessiamarcolini — APPROVED** [review]

File: src/histolab/slide.py [hunk 18]
```diff
@@ -135,6 +145,59 @@ def level_dimensions(self, level: int = 0) -> Tuple[int, int]:
...
+        magnification factor : str
+            Magnification factor at speficied level
+
+        Raises
+        ------
+        LevelError
+            If the specified level is not available.
+        SlidePropertyError
+            If the slide's native magnification is not available in the file
+            metadata.
```
Comment (ernestoarbitrio) [inline src/histolab/slide.py:170]: I think here we should add an `or` in the sense that we raise that execption in 2 cases and we mentioned only 1. So I would include 
```
or if Downsample factor for given level is not available.
```

**Review by ernestoarbitrio — APPROVED** [review]
LGTM but let's see if you can address the minor comment I just wrote

**Review by ernestoarbitrio — APPROVED** [review]
LGTM well done 👯

**Pull request merged**